yt_dlp/extractor/espn.py

```diff
@@ -279,3 +278,133 @@ def _real_extract(self, url):
+        if 'authorization' not in headers:
+            headers['authorization'] = 'Bearer ' + self._API_KEY
+        return self._download_json(
+            'https://espn.api.edge.bamgrid.com/' + path,
+            video_id, data=data, headers=headers)
+
+    def _payload_to_string_data(self, payload):
+        return '&'.join(['%s=%s' % (key, compat_urllib_parse_quote_plus(payload[key])) for key in payload]).encode('utf-8')
+
+    def _payload_to_json_data(self, payload):
+        return json.dumps(payload).encode('utf-8')
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        video_data = self._download_json(
+            'https://watch-cdn.product.api.espn.com/api/product/v3/watchespn/web/playback/event?id=%s' % video_id, video_id)['playbackState']
+        title = video_data['name']
+
+        # ESPN+ subscription required, through cookies
+        if video_data.get('sourceId') == 'ESPN_DTC':
+            cookies = self._get_cookies(url)
+            try:
+                id_token = cookies['ESPN-ONESITE.WEB-PROD.token'].value.split('|')[1]
+            except KeyError:
+                raise ExtractorError('This is an ESPN+ video, which requires cookies. Use --cookies or --cookiesfrombrowser')
+
+            assertion = self._call_bamgrid_api(
+                'devices',
+                video_id,
+                data=self._payload_to_json_data({
+                    'deviceFamily': 'android',
+                    'applicationRuntime': 'android',
+                    'deviceProfile': 'tv',
+                    'attributes': {},
+                }),
+                headers={'content-type': 'application/json; charset=UTF-8'}
+            )['assertion']
+
+            token = self._call_bamgrid_api(
+                'token',
+                video_id,
+                data=self._payload_to_string_data({
+                    'subject_token': assertion,
+                    'subject_token_type': 'urn:bamtech:params:oauth:token-type:device',
+                    'platform': 'android',
+                    'grant_type': 'urn:ietf:params:oauth:grant-type:token-exchange'})
+            )['access_token']
+
+            assertion = self._call_bamgrid_api(
+                'accounts/grant',
+                video_id,
+                data=self._payload_to_json_data({'id_token': id_token}),
+                headers={
+                    'authorization': token,
+                    'content-type': 'application/json; charset=UTF-8'}
+            )['assertion']
+
+            token = self._call_bamgrid_api(
+                'token',
+                video_id,
+                data=self._payload_to_string_data({
+                    'subject_token': assertion,
+                    'subject_token_type': 'urn:bamtech:params:oauth:token-type:account',
+                    'platform': 'android',
+                    'grant_type': 'urn:ietf:params:oauth:grant-type:token-exchange'})
+            )['access_token']
```

Comment:
does all this really have to be done for each video? or it can be done once and the token saved

Reply:
Not 100% sure here, although I can try removing the first couple calls and seeing what happens. Actually, when working on this earlier, I was manually hardcoding the value for `token` that gets passed into the accounts/grant call, so maybe that could be cached. However, I know that these tokens expire after around a few hours. I wouldn't want to save a token and then when a user goes to download a video a few hours later, the token be expired. Plus, I'm not sure how to implement such caching logic to begin with.

I will say that there doesn't seem to be any _consequence_ to running through these steps every time, other than having more lines in the code. What I mean is that I can test like a dozen URLs in a minute and I still get the expected responses.

Reply:
Ah fair enough. I was considering including it in a `_real_initialize`, but then it would attempt to do all this stuff even if the user was passing in URLs that are authenticated TV providers and not for ESPN+. This is a tricky scenario where essentially we should have two different extractors labeled two different ways, but as they both follow the same URL schema I'm not sure how feasible this is.